File: backend/engineering/network_crossings.py
"""Drawing-only bridge geometry. Physical bus identity decides connectivity."""
from bisect import bisect_left

EPS = .001


def _number(value):
    return f'{value:.3f}'.rstrip('0').rstrip('.') if value else '0'


def _point(x, y):
    return f'{_number(x)} {_number(y)}'


def with_wire_crossings(scene):
    wires = []
    for bus in scene['buses']:
        ends = [branch['points'][-1] for branch in bus['branches']]
        if not ends:
            continue
        wires.append({'busId': bus['id'], 'branch': -1, 'points': [
            {'x': ends[0]['x'], 'y': min(p['y'] for p in ends)}, {'x': ends[0]['x'], 'y': max(p['y'] for p in ends)}]})
        wires.extend({'busId': bus['id'], 'branch': i, 'points': branch['points']} for i, branch in enumerate(bus['branches']))
    segments = []
    for wire_index, wire in enumerate(wires):
        for i, (a, b) in enumerate(zip(wire['points'], wire['points'][1:])):
            vertical = abs(a['x'] - b['x']) < EPS
            if not vertical and abs(a['y'] - b['y']) >= EPS:
                continue
            axis = 'y' if vertical else 'x'
            lo, hi = sorted((a[axis], b[axis]))
            if hi - lo > EPS:
                segments.append({'wire': wire_index, 'index': i, 'vertical': vertical, 'fixed': a['x' if vertical else 'y'], 'lo': lo, 'hi': hi})
    verticals = sorted((i for i, s in enumerate(segments) if s['vertical']), key=lambda i: segments[i]['fixed'])
    xs = [segments[i]['fixed'] for i in verticals]
    jumps = {}
    for h_index, h in enumerate(segments):
        if h['vertical']:
            continue
        i = bisect_left(xs, h['lo'] - EPS)
        while i < len(verticals) and xs[i] <= h['hi'] + EPS:
            v_index = verticals[i]
            v = segments[v_index]
            i += 1
            if wires[h['wire']]['busId'] == wires[v['wire']]['busId'] or not v['lo'] - EPS <= h['fixed'] <= v['hi'] + EPS:
                continue
            h_room = min(v['fixed'] - h['lo'], h['hi'] - v['fixed'])
            v_room = min(h['fixed'] - v['lo'], v['hi'] - h['fixed'])
            target = h_index if h_room > 3 else v_index if v_room > 3 else None
            if target is not None:
                jumps.setdefault(target, []).append(h['fixed'] if segments[target]['vertical'] else v['fixed'])
    gaps, bridges, emitted = {}, [], set()
    for segment_index, values in jumps.items():
        segment = segments[segment_index]
        ranges = []
        for at in sorted(set(values)):
            radius = min(6, at - segment['lo'] - 1, segment['hi'] - at - 1)
            start, end = at - radius, at + radius
            if ranges and start <= ranges[-1][1] + 2:
                ranges[-1][1] = max(ranges[-1][1], end)
            else:
                ranges.append([start, end])
        gaps[segment['wire'], segment['index']] = ranges
        for start, end in ranges:
            wire = wires[segment['wire']]
            key = wire['busId'], segment['vertical'], _number(segment['fixed']), _number(start), _number(end)
            if key in emitted:
                continue
            emitted.add(key)
            half = (end - start) / 2
            fixed = segment['fixed']
            path = (f'M {_point(fixed, start)} A 6 {_number(half)} 0 0 1 {_point(fixed, end)}' if segment['vertical'] else
                    f'M {_point(start, fixed)} A {_number(half)} 6 0 0 1 {_point(end, fixed)}')
            bridges.append({'busId': wire['busId'], 'branch': wire['branch'], 'path': path})
    paths = {}
    for index, wire in enumerate(wires):
        commands = [f"M {_point(wire['points'][0]['x'], wire['points'][0]['y'])}"]
        for i, (a, b) in enumerate(zip(wire['points'], wire['points'][1:])):
            vertical = abs(a['x'] - b['x']) < EPS
            forward = b['y'] >= a['y'] if vertical else b['x'] >= a['x']
            ranges = gaps.get((index, i), [])
            for r in ranges if forward else reversed(ranges):
                start, end = r if forward else reversed(r)
                commands.extend([f"L {_point(a['x'], start) if vertical else _point(start, a['y'])}",
                                 f"M {_point(a['x'], end) if vertical else _point(end, a['y'])}"])
            commands.append(f"L {_point(b['x'], b['y'])}")
        paths[wire['busId'], wire['branch']] = ' '.join(commands)
    return {**scene, 'crossingVersion': 1, 'wireBridges': bridges, 'buses': [
        {**bus, 'displayPath': paths.get((bus['id'], -1), bus.get('path', '')), 'branches': [
            {**branch, 'displayPath': paths.get((bus['id'], i), branch.get('path', ''))} for i, branch in enumerate(bus['branches'])]}
        for bus in scene['buses']]}

File: frontend/src/lib/network-crossings.ts
import type { NetworkScene, ScenePoint } from './network-scene';

type Wire = {busId: string; branch: number; points: ScenePoint[]};
type Segment = {wire: number; index: number; vertical: boolean; fixed: number; lo: number; hi: number};
export type WireBridge = {busId: string; branch: number; path: string};
const EPS = .001;
const number = (value: number) => String(Math.round(value * 1000) / 1000);
const point = (x: number, y: number) => `${number(x)} ${number(y)}`;

/** Separate physical buses cross with a jump; same-bus junctions stay connected. */
export function withWireCrossings<T extends Pick<NetworkScene, 'buses'>>(scene: T): T & {wireBridges: WireBridge[]; crossingVersion: number} {
  const wires: Wire[] = [];
  for (const bus of scene.buses) {
    const ends = bus.branches.map(b => b.points.at(-1)!);
    if (!ends.length) continue;
    wires.push({busId: bus.id, branch: -1, points: [{x: ends[0].x, y: Math.min(...ends.map(p => p.y))}, {x: ends[0].x, y: Math.max(...ends.map(p => p.y))}]});
    bus.branches.forEach((b, branch) => wires.push({busId: bus.id, branch, points: b.points}));
  }
  const segments: Segment[] = [];
  wires.forEach((wire, index) => wire.points.slice(1).forEach((b, i) => {
    const a = wire.points[i], vertical = Math.abs(a.x - b.x) < EPS;
    if (!vertical && Math.abs(a.y - b.y) >= EPS) return;
    const lo = Math.min(vertical ? a.y : a.x, vertical ? b.y : b.x);
    const hi = Math.max(vertical ? a.y : a.x, vertical ? b.y : b.x);
    if (hi - lo > EPS) segments.push({wire:index, index:i, vertical, fixed:vertical?a.x:a.y, lo, hi});
  }));
  const verticals = segments.filter(s => s.vertical).sort((a,b) => a.fixed-b.fixed);
  const jumps = new Map<Segment, number[]>();
  for (const h of segments.filter(s => !s.vertical)) {
    // Binary search avoids comparing every pair in large, mostly separate clusters.
    let lo=0, hi=verticals.length;
    while (lo<hi) {const mid=(lo+hi)>>>1;if(verticals[mid].fixed<h.lo-EPS)lo=mid+1;else hi=mid;}
    for (let i=lo;i<verticals.length&&verticals[i].fixed<=h.hi+EPS;i++) {
      const v=verticals[i];
      if(wires[h.wire].busId===wires[v.wire].busId || h.fixed<v.lo-EPS || h.fixed>v.hi+EPS)continue;
      const hRoom=Math.min(v.fixed-h.lo,h.hi-v.fixed), vRoom=Math.min(h.fixed-v.lo,v.hi-h.fixed);
      // Prefer a horizontal jump. A T-touch can instead jump the vertical wire.
      const target=hRoom>3?h:vRoom>3?v:null;
      if(!target)continue; // Coincident endpoints are not an interior crossing.
      const values=jumps.get(target)??[];values.push(target.vertical?h.fixed:v.fixed);jumps.set(target,values);
    }
  }
  const gaps = new Map<string, Array<[number,number]>>();
  const wireBridges: WireBridge[] = [];
  const emitted=new Set<string>();
  for(const [segment, values] of jumps) {
    const ranges:Array<[number,number]>=[];
    for(const at of [...new Set(values)].sort((a,b)=>a-b)) {
      const radius=Math.min(6,at-segment.lo-1,segment.hi-at-1);
      const start=at-radius, end=at+radius, previous=ranges.at(-1);
      if(previous&&start<=previous[1]+2)previous[1]=Math.max(previous[1],end);else ranges.push([start,end]);
    }
    gaps.set(`${segment.wire}:${segment.index}`,ranges);
    for(const [start,end] of ranges) {
      const wire=wires[segment.wire], key=`${wire.busId}:${segment.vertical}:${number(segment.fixed)}:${number(start)}:${number(end)}`;
      if(emitted.has(key))continue;emitted.add(key);
      const half=(end-start)/2;
      const d=segment.vertical
        ? `M ${point(segment.fixed,start)} A 6 ${number(half)} 0 0 1 ${point(segment.fixed,end)}`
        : `M ${point(start,segment.fixed)} A ${number(half)} 6 0 0 1 ${point(end,segment.fixed)}`;
      wireBridges.push({busId:wire.busId,branch:wire.branch,path:d});
    }
  }
  const paths=new Map<string,string>();
  wires.forEach((wire,index)=>{
    const commands=[`M ${point(wire.points[0].x,wire.points[0].y)}`];
    wire.points.slice(1).forEach((b,i)=>{
      const a=wire.points[i],vertical=Math.abs(a.x-b.x)<EPS;
      const forward=vertical?b.y>=a.y:b.x>=a.x;
      const ranges=gaps.get(`${index}:${i}`)??[];
      for(const range of forward?ranges:[...ranges].reverse()) {
        const [start,end]=forward?range:[range[1],range[0]];
        commands.push(`L ${vertical?point(a.x,start):point(start,a.y)}`,`M ${vertical?point(a.x,end):point(end,a.y)}`);
      }
      commands.push(`L ${point(b.x,b.y)}`);
    });
    paths.set(`${wire.busId}:${wire.branch}`,commands.join(' '));
  });
  return {...scene,crossingVersion:1,wireBridges,buses:scene.buses.map(bus=>({...bus,
    displayPath:paths.get(`${bus.id}:-1`)??bus.path,
    branches:bus.branches.map((branch,index)=>({...branch,displayPath:paths.get(`${bus.id}:${index}`)??branch.path}))}))};
}
